pkg/bruteforcer/brute_forcer.go

```diff
@@ -0,0 +1,187 @@
+	ctx, err := b.initFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	if b.checkFunc(ctx, b.initialData) {
+		return NewUniqueUnorderedCombination(0), nil
+	}
+
+	totalBitLength := uint64(len(b.initialData)) * 8
+	if maxDistance > totalBitLength {
+		maxDistance = totalBitLength
+	}
+
+	for distance := uint64(1); distance <= maxDistance; distance++ {
+		if totalBitLength < distance {
+			// no combinations possible
+			break
+		}
+		amountOfCombinationsBigInt := big.NewInt(1).Binomial(int64(totalBitLength), int64(distance))
+
+		// Current algorithm uses an uint64 as an iterator through combinations,
+		// thus we do not support more than math.MaxInt64 combinations.
+		//
+		// And amount of
+		//
+		// ">=" instead of ">" just for defense.
+		if amountOfCombinationsBigInt.Cmp(big.NewInt(math.MaxInt64)) >= 0 {
+			return nil, fmt.Errorf("distance is too high (amount of combinations causes uint64 overflow)")
+		}
+		amountOfCombinations := amountOfCombinationsBigInt.Uint64()
+
+		// Calculating:
+		// * concurrencyFactor -- how many concurrent goroutines are permitted.
+		// * and combinationsPiece -- how many combinations should one goroutine handle.
+
+		concurrencyFactor := runtime.GOMAXPROCS(0)
+		if uint64(concurrencyFactor) > amountOfCombinations/minIterationsPerCPU {
+			concurrencyFactor = int(amountOfCombinations / minIterationsPerCPU)
+			if concurrencyFactor < 1 {
+				concurrencyFactor = 1
+			}
+		}
+		if maxConcurrency > 0 && uint(concurrencyFactor) > maxConcurrency {
+			concurrencyFactor = int(maxConcurrency)
+		}
+		combinationsPiece := amountOfCombinations / uint64(concurrencyFactor)
+
+		// Brute force!
+
+		var locker sync.Mutex
+		var resultData UniqueUnorderedCombination
+		var wg sync.WaitGroup
+		errChan := make(chan error, concurrencyFactor)
+
+		for i := 0; i < concurrencyFactor; i++ {
+			combinationIDStart := uint64(i) * combinationsPiece
+			combinationIDEnd := uint64(i+1) * combinationsPiece
+			if i == concurrencyFactor-1 {
+				combinationIDEnd = amountOfCombinations
+			}
+
+			wg.Add(1)
+			go func(combinationIDStart, combinationIDEnd uint64, errChan chan<- error) {
+				defer wg.Done()
+				dataCopy := make([]byte, len(b.initialData))
+				copy(dataCopy, b.initialData)
+
+				iterator := NewUniqueUnorderedCombinationIterator(distance, int64(totalBitLength)-1)
+				iterator.SetCombinationID(combinationIDStart)
+
+				bfctx, err := b.initFunc()
+				if err != nil {
+					errChan <- fmt.Errorf("brute forcer initFunc error for range (%v, %v): %v", combinationIDStart, combinationIDEnd, err)
+					return
+				}
+
+				combinations := combinationIDEnd - combinationIDStart
+				for i := uint64(0); ; {
+					if resultData != nil {
+						// we do not need atomic in the condition above, because
+						// few false negatives will not affect anything much.
+						return
+					}
+
+					if b.try(bfctx, dataCopy, iterator) {
+						locker.Lock()
+						resultData = iterator.GetCombination()
```

Comment:
nit

it would be nice to notify other goroutines to stop

Reply:
We do it above in:
```
					if resultData != nil {
						// we do not need atomic in the condition above, because
						// few false negatives will not affect anything much.
						return
					}
```
In the worst case scenario other routines would do like 2 extra iterations.